src/sidebar/components/thread-list-omega.js

```diff
@@ -0,0 +1,246 @@
+import { createElement } from 'preact';
+import { useEffect, useState } from 'preact/hooks';
+import propTypes from 'prop-types';
+import debounce from 'lodash.debounce';
+import shallowEqual from 'shallowequal';
+
+import events from '../events';
+import useStore from '../store/use-store';
+import { isHighlight, isReply } from '../util/annotation-metadata';
+import { getElementHeightWithMargins } from '../util/dom';
+import { withServices } from '../util/service-context';
+
+import Thread from './thread';
+
+// When we don't have a real measurement of a thread card's height (yet)
+// from the browser, use this as an approximate value, in pixels.
+const THREAD_DEFAULT_HEIGHT = 200;
+
+/**
+ * Calculate the set of `thread`s that should be rendered by estimating which
+ * of the threads are within or near the viewport.
+ *
+ * @param {Thread[]} threads - List of threads in the order they appear
+ * @param {Object} threadHeights - Map of thread ID to measured height
+ * @param {number} scrollPos - Vertical scroll offset of scrollable container
+ * @param {number} windowHeight -
+ *   Height of the visible area of the scrollable container.
+ */
+function calculateVisibleThreads(
+  threads,
+  threadHeights,
+  scrollPos,
+  windowHeight
```

Comment:
Note this function now takes the _scroll position_ and _window height_ as numeric arguments rather than being given DOM elements to query. This makes it "safe" to use from within the body of a render function.